data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroSchemaParserFromTabularFormat.java

```diff
@@ -0,0 +1,194 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class AvroSchemaParserFromTabularFormat {
+
+    private static final String END_SCHEMA_STRING = "]}";
+    private static final Pattern pattern = Pattern.compile("\\(([^()]*|)*\\)");
+
+    static Schema generateSchemaFromTabularString(final String inputString) throws IOException {
+        String recordSchemaOutputString = inputString;
+        recordSchemaOutputString = recordSchemaOutputString.trim();
+
+        final String tableName = extractTableName(recordSchemaOutputString.split("\\s+"));
+        recordSchemaOutputString = getStringFromParanthesis(recordSchemaOutputString);
+        recordSchemaOutputString = removeSpaceAndReplaceParanthesis(recordSchemaOutputString);
+
+        final StringBuilder mainSchemaBuilder = new StringBuilder();
+        final String baseSchemaStr = "{\"type\":\"record\",\"name\":\"" + tableName + "\",\"fields\":[";
+        mainSchemaBuilder.append(baseSchemaStr);
+
+        iterateRecursively(mainSchemaBuilder, recordSchemaOutputString, false, tableName, 0);
+
+        mainSchemaBuilder.append(END_SCHEMA_STRING);
+
+        return new org.apache.avro.Schema.Parser().parse(mainSchemaBuilder.toString());
+    }
+
+    private static String removeSpaceAndReplaceParanthesis(String recordSchemaOutputString) {
+        recordSchemaOutputString = recordSchemaOutputString.replaceAll("\\(", "");
+        recordSchemaOutputString = recordSchemaOutputString.replaceAll("\\)", "");
+
+        recordSchemaOutputString = recordSchemaOutputString.trim();
+
+        recordSchemaOutputString = recordSchemaOutputString.replaceAll("\\n", "");
+        recordSchemaOutputString = recordSchemaOutputString.replaceAll(",\\s+", ",");
+        recordSchemaOutputString = recordSchemaOutputString.replaceAll(">\\s+", ">");
+        recordSchemaOutputString = recordSchemaOutputString.replaceAll("<\\s+", "<");
+        return recordSchemaOutputString;
+    }
+
+    private static String getStringFromParanthesis(String recordSchemaOutputString) {
+        final Matcher matcher = pattern.matcher(recordSchemaOutputString);
+        if (matcher.find()) {
+            recordSchemaOutputString = matcher.group(0);
+        }
+        return recordSchemaOutputString;
+    }
+
+    private static String extractTableName(final String[] words) throws IOException {
+        String tableName = null;
+        if (words.length >= 2) {
+            tableName = words[1];
+        } else {
+            throw new IOException("Invalid schema string.");
+        }
+        return tableName;
+    }
+
+    private static String buildBaseSchemaString(final String part) {
+        return "{\"type\":\"record\",\"name\":\"" + part + "\",\"fields\":[";
+    }
+
+    private static String buildSchemaStringForArr() {
+        return "{\"type\":\"array\", \"items\":\"string\"}";
+    }
+
+    public static void iterateRecursively(final StringBuilder mainSchemaBuilder, final String recordSchema,
```

Comment:
This code appears to be taking a SQL-like schema and then converting it into an Avro schema. Right?

Why not just accept the Avro schema in the pipeline configuration?

This appears to be introducing an undefined language and there may be many different edge cases that we are not accounting for.

If this is using a well defined language, can we perform a model mapping to make sure it is valid? Say for example, it is a Postgresql DDL, can we parse the DDL using a Postgresql library into a Java model? Then we can perform a more accurate mapping.

Reply:
Yes @dlvenable , This is Raj's suggestion to support glue like schema and converting it onto avro schema, We are already accepting Avro schema in pipeline YAML itself, 
As of now there is no library to do mapping between glue schema like structure to Avro/Parquet schema as there are nested and logical types in Avro/Parquet also.

Reply:
@dlvenable , Shared over mail